core/templates/rid_owner.h

```diff
@@ -69,42 +69,47 @@ class RID_AllocBase {
 
 template <class T, bool THREAD_SAFE = false>
 class RID_Alloc : public RID_AllocBase {
-	T **chunks = nullptr;
+	struct Chunk {
+		T data;
+		uint32_t validator;
+	};
+	Chunk **chunks = nullptr;
 	uint32_t **free_list_chunks = nullptr;
-	uint32_t **validator_chunks = nullptr;
 
 	uint32_t elements_in_chunk;
 	uint32_t max_alloc = 0;
 	uint32_t alloc_count = 0;
+	uint32_t chunk_limit = 0;
 
 	const char *description = nullptr;
 
-	mutable SpinLock spin_lock;
+	mutable Mutex mutex;
 
 	_FORCE_INLINE_ RID _allocate_rid() {
 		if (THREAD_SAFE) {
-			spin_lock.lock();
+			mutex.lock();
 		}
 
 		if (alloc_count == max_alloc) {
 			//allocate a new chunk
 			uint32_t chunk_count = alloc_count == 0 ? 0 : (max_alloc / elements_in_chunk);
+			ERR_FAIL_COND_V(THREAD_SAFE && chunk_count == chunk_limit, RID());
```

Comment:
This will leave the mutex permanently locked

Reply:
Oh you rock, nicely spotted, thanks!